File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/SecurityPreservationProof.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.security.SecurityInvariantRegistry;
import java.util.*;

/**
 * Records that a semantic rewrite preserved the security contract of the input
 * plan.
 */
public record SecurityPreservationProof(List<String> before, List<String> after, List<String> missing,
		String beforeFingerprint, String afterFingerprint) {
	public SecurityPreservationProof {
		before = before == null ? List.of() : List.copyOf(before);
		after = after == null ? List.of() : List.copyOf(after);
		missing = missing == null ? List.of() : List.copyOf(missing);
	}

	public boolean isSatisfied() {
		return missing.isEmpty() && before.stream().sorted().toList().equals(after.stream().sorted().toList());
	}

	public static SecurityPreservationProof create(SemanticPlan beforePlan, SemanticPlan afterPlan) {
		Objects.requireNonNull(beforePlan);
		Objects.requireNonNull(afterPlan);
		for (var id : beforePlan.effectiveSecurityInvariants())
			if (!SecurityInvariantRegistry.contains(id))
				throw new IllegalStateException("Unknown security invariant '" + id + "' in source plan.");
		for (var id : afterPlan.effectiveSecurityInvariants())
			if (!SecurityInvariantRegistry.contains(id))
				throw new IllegalStateException("Unknown security invariant '" + id + "' in rewritten plan.");
		var afterSet = new HashSet<>(afterPlan.effectiveSecurityInvariants());
		var missing = beforePlan.effectiveSecurityInvariants().stream().filter(id -> !afterSet.contains(id)).distinct()
				.sorted().toList();
		var proof = new SecurityPreservationProof(beforePlan.effectiveSecurityInvariants(),
				afterPlan.effectiveSecurityInvariants(), missing, SemanticPlanFingerprint.create(beforePlan),
				SemanticPlanFingerprint.create(afterPlan));
		if (!proof.isSatisfied())
			throw new IllegalStateException(
					"Security-preserving rewrite rejected. Missing invariants: " + String.join(", ", missing) + ".");
		return proof;
	}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/SecurityPreservationProof.cs
using Foundgine.Core.Semantic.Security;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Records that a semantic rewrite preserved the security contract of the
/// input plan. This is a rewrite proof, not a claim that the provider is
/// correct or that authorization has been granted.
/// </summary>
public sealed record SecurityPreservationProof(
    IReadOnlyList<string> Before,
    IReadOnlyList<string> After,
    IReadOnlyList<string> Missing,
    string BeforeFingerprint,
    string AfterFingerprint)
{
    public bool IsSatisfied => Missing.Count == 0 &&
                               Before.OrderBy(x => x, StringComparer.Ordinal).SequenceEqual(
                                   After.OrderBy(x => x, StringComparer.Ordinal), StringComparer.Ordinal);

    public static SecurityPreservationProof Create(SemanticPlan before, SemanticPlan after)
    {
        ArgumentNullException.ThrowIfNull(before);
        ArgumentNullException.ThrowIfNull(after);

        foreach (var id in before.EffectiveSecurityInvariants)
            if (!SecurityInvariantRegistry.Contains(id))
                throw new InvalidOperationException($"Unknown security invariant '{id}' in source plan.");

        foreach (var id in after.EffectiveSecurityInvariants)
            if (!SecurityInvariantRegistry.Contains(id))
                throw new InvalidOperationException($"Unknown security invariant '{id}' in rewritten plan.");

        var afterSet = after.EffectiveSecurityInvariants.ToHashSet(StringComparer.Ordinal);
        var missing = before.EffectiveSecurityInvariants
            .Where(id => !afterSet.Contains(id))
            .Distinct(StringComparer.Ordinal)
            .OrderBy(id => id, StringComparer.Ordinal)
            .ToArray();

        var proof = new SecurityPreservationProof(
            before.EffectiveSecurityInvariants.ToArray(),
            after.EffectiveSecurityInvariants.ToArray(),
            missing,
            SemanticPlanFingerprint.Create(before),
            SemanticPlanFingerprint.Create(after));

        if (!proof.IsSatisfied)
            throw new InvalidOperationException(
                $"Security-preserving rewrite rejected. Missing invariants: {string.Join(", ", missing)}.");

        return proof;
    }
}